portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java

```diff
@@ -18,44 +18,107 @@
-        String userId = userService.getCurrentUser().getUserId();
 
-        String nickname = getUserNickname(userId);
-        if (nickname == null) nickname = "Anonymous";
+        JsonObject json = new JsonObject();
 
-        response.getWriter().println("{ \"nickname\" : \"" + nickname + "\" }");
+        if (!userService.isUserLoggedIn()) {
+            String loginUrl = userService.createLoginURL("/");
+            json.addProperty("loginUrl", loginUrl);
+            json.addProperty("userLoggedIn", false);
+        } else {
+            String nickname = getCurrentUserNickname();
+            if (nickname == null) {
+                nickname = "Nameless-individual";
+            }
+            json.addProperty("nickname", nickname);
+            json.addProperty("userLoggedIn", true);
+        }
+
+        response.getWriter().println(json.toString());
     }
 
+    // doPost implements settings.html post request.
+    // Associates settings input with user account entity in datastore.
+    // Relevant entities contain public profile information only.
     @Override
     public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-        //nothing
-    }
+        String nickname = request.getParameter("nickname-input");
+
+        if (nickname == null) {
+            System.out.println("invalid nickname requested");
+            return;
+        }
+
+        nickname = escapeSpecialChars(nickname);
 
-    private String getUserNickname(String id) {
         DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-        Query query =
-            new Query("UserInfo")
-                .setFilter(new Query.FilterPredicate("id", Query.FilterOperator.EQUAL, id));
-        PreparedQuery results = datastore.prepare(query);
-        Entity entity = results.asSingleEntity();
-        if (entity == null) {
+        UserService userService = UserServiceFactory.getUserService();
+        String userEmail = userService.getCurrentUser().getEmail();
+
+        Entity entity = new Entity("userPublicProfile", userEmail);
+        entity.setProperty("nickname", nickname);
+
+        datastore.put(entity);
+
+        response.sendRedirect("/settings.html");
+    }
+
+    // https://stackoverflow.com/questions/3844595/
+
+  private String escapeSpecialChars(String str) {
+    StringBuilder builder = new StringBuilder();
+    for( char c : str.toCharArray() )
+    {
+        if( c == '\'' )
+            builder.append( "\\'" );
+        else if ( c == '\"' )
+            builder.append( "\\\"" );
+        else if( c == '\r' )
+            builder.append( "\\r" );
+        else if( c == '\n' )
+            builder.append( "\\n" );
+        else if( c == '\t' )
+            builder.append( "\\t" );
+        else if( c < 32 || c >= 127 )
+            builder.append( String.format( "\\u%04x", (int)c ) );
+        else
+            builder.append( c );
+    }
```

Comment:
same here as comment above.

Also, this is totally up to you, but when you have repeated code like this, you can break out into a utils file, so you can use it from multiple classes but you only have to write it once.